deepchem/models/losses.py

```diff
@@ -465,6 +468,119 @@ def loss(inputs):
         return loss
 
 
+class GroverPretrainLoss(Loss):
+    """GroverPretrainLoss
+
+    The Grover Pretraining consists learning of atom embeddings and bond embeddings for
+    a molecule. To this end, the learning consists of three tasks:
+        1. Learning of atom vocabulary from atom embeddings and bond embeddings
+        2. Learning of bond vocabulary from atom embeddings and bond embeddings
+        3. Learning to predict functional groups from atom embedings readout and bond embeddings readout
+    The loss function accepts atom vocabulary labels, bond vocabulary labels and functional group
+    predictions produced by Grover model during pretraining as a dictionary and applies negative
+    log-likelihood loss for atom vocabulary and bond vocabulary predictions and Binary Cross Entropy
+    loss for functional group prediction and sums these to get overall loss.
+    """
+
+    def _create_pytorch_loss(self):
+        import torch.nn as nn
+
+        def loss(preds,
+                 targets,
+                 weights: Optional[List[Sequence]] = None,
+                 dist_coff=0.1):
+            """
+            Parameters
+            -----------
+            preds: torch.Tensor
+                predictions of a grover pretraining model
+            targets: torch.Tensor
+                targets (real values)
+            dist_coff: float, default 0.1
+                loss metric
+
+            Returns
+            -------
+            loss: torch.Tensor
+                loss value
+            """
+            av_task_loss = nn.NLLLoss(ignore_index=0,
+                                      reduction="mean")  # same for av and bv
+            fg_task_loss = nn.BCEWithLogitsLoss(reduction="mean")
+            av_task_dist_loss = nn.MSELoss(reduction="mean")
+            fg_task_dist_loss = nn.MSELoss(reduction="mean")
+
+            sigmoid = nn.Sigmoid()
+
+            av_atom_loss = av_task_loss(preds['av_task'][0], targets["av_task"])
```

Comment:
What is the expected structure of `preds`? Why do we expect it to have a field 'av_task'? We can do one round of simplification here to directly accept atom losses, bond loses, etc as the arguments to this loss function

Reply:
There are 3 sets of predictions from GroverPretrain model and they are encoded in the dict `preds`. It is a dict with keys `av_task`, `bv_task` and `fg_task`.